doc/developer/design/20230512_mir_cost_model.md

```diff
@@ -0,0 +1,149 @@
+
+A *cost model* tries to predict a program’s behavior given information about its inputs. Given a program and some contextual information, a cost model produces an *estimated cost* for that query. Optimizers use cost models to make better informed decisions.
+
+Our cost model should be in terms of the dimensions that are relevant to us:
+
+- Latency (rehydration time; latency per update)
+- Storage (memory footprint/implicit storage)
+- Scalability (to what degree can this query be spread across workers)
+
+Our cost model should be *executable* and *accurate:*
+
+- Our cost model will be code that runs on actual MIR, producing a symbolic expression that represents the cost of the query along various dimensions. A user can use `EXPLAIN WITH(cost)`  to predict costs.
+- Our cost model should be tested on a variety of loads to ensure that our estimates aren’t total nonsense.
+
+# Motivation
+[motivation]: #motivation
+
+## Goals
+
+- Reify what we learn from various benchmarks/experiments.
+    - https://github.com/MaterializeInc/materialize/issues/17883
+    - https://github.com/MaterializeInc/materialize/issues/16511
+    - https://github.com/MaterializeInc/materialize/issues/15244
+    - https://github.com/MaterializeInc/materialize/issues/13140
+    - https://github.com/MaterializeInc/materialize/issues/12069
+    - https://github.com/MaterializeInc/materialize/issues/9157
+- Offer visibility/predicability.
+    - https://github.com/MaterializeInc/materialize/issues/14663
+    - https://github.com/MaterializeInc/materialize/issues/13046
+- Open opportunities for principled/well-informed choices in the optimizer.
+
+## Non-goals
+
+- Identify all opportunities to use the cost model in the optimizer.
+- Refactor the optimizer to be focused on the cost model.
+- Implement dynamic re-optimization based on changes in input statistics.
+- Collect all relevant statistics in storage and make them available to the cost model.
+
+# Explanation
+[explanation]: #explanation
+
+## Proposal
+
+A cost model is an abstract interpretation of the query: rather than running the query on relations and getting new relation out, we run the query on a symbolic description of its input relations and get a symbolic description of the result out.
+
+The general approach can be the same as `mz_transform::Typecheck`: recursively walk the `MirRelationExpr` with some contextual information, returning a symbolic term characterizing the query.
+
+### Context information
+
+Traditional databases track cardinality data and statistics about various columns. We will be interested in other statistics, too. Some possible examples:
+
+- How frequently does data arrive?
+- Is the given input physically monotonic or append-only? (I.e., there are no retractions.)
```

Comment:
Right now, this is tricky to fully observe in MIR, since MIR planning is not parameterized by whether we are in a single-time context or not. Rather, we observe only source monotonicity at MIR and defer considering single-time monotonicity to a plan refinement at LIR. This creates a tension between the `MonotonicFlag` MIR analysis and what is then done in LIR (more discussion [here](https://github.com/vmarcos/materialize/blob/stabilize_monotonic_select/doc/developer/design/20230421_stabilize_monotonic_select.md#relationship-to-monotonicity-analysis-in-mir-and-explain)). This tension implies that we would need to reanalyze, in total or in part, monotonicity at the `Plan` level, to consider single-time contexts (more discussion [here](https://github.com/vmarcos/materialize/blob/stabilize_monotonic_select/doc/developer/design/20230421_stabilize_monotonic_select.md#improve-refine_single_time_dataflow-lir-based-refinement-materializeincmaterialize18732)).

Taking a step back, there is some need to consider here how to derive an MIR-level cost estimate for a one-shot `SELECT` if some of the decisions affecting query processing time end up being made only at LIR.